src/interface/swarm_device_context.hpp

```diff
@@ -0,0 +1,108 @@
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+#define INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+
+#include "coordinates/coordinates.hpp"
+
+namespace parthenon {
+
+class SwarmDeviceContext {
+ public:
+  KOKKOS_FUNCTION
+  bool IsActive(int n) const { return mask_(n); }
+
+  KOKKOS_FUNCTION
+  bool IsOnCurrentMeshBlock(int n) const { return blockIndex_(n) == this_block_; }
+
+  KOKKOS_FUNCTION
+  void MarkParticleForRemoval(int n) const { marked_for_removal_(n) = true; }
+
+  KOKKOS_FUNCTION
+  bool IsMarkedForRemoval(const int n) const { return marked_for_removal_(n); }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetNeighborBlockIndex(const int &n, const double &x, const double &y,
+                            const double &z, bool &is_on_current_mesh_block) const {
+    int i = static_cast<int>(std::floor((x - x_min_) / ((x_max_ - x_min_) / 2.))) + 1;
+    int j = static_cast<int>(std::floor((y - y_min_) / ((y_max_ - y_min_) / 2.))) + 1;
+    int k = static_cast<int>(std::floor((z - z_min_) / ((z_max_ - z_min_) / 2.))) + 1;
+
+    // Something went wrong
+    if (i < 0 || i > 3 || ((j < 0 || j > 3) && ndim_ > 1) ||
+        ((k < 0 || k > 3) && ndim_ > 2)) {
+      PARTHENON_FAIL("Particle neighbor indices out of bounds");
+    }
+
+    // Ignore k,j indices as necessary based on problem dimension
+    if (ndim_ == 1) {
+      blockIndex_(n) = neighborIndices_(0, 0, i);
+    } else if (ndim_ == 2) {
+      blockIndex_(n) = neighborIndices_(0, j, i);
+    } else {
+      blockIndex_(n) = neighborIndices_(k, j, i);
+    }
+
+    is_on_current_mesh_block = (blockIndex_(n) == this_block_);
+
+    return blockIndex_(n);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetMyRank() const { return my_rank_; }
+
+  KOKKOS_INLINE_FUNCTION
+  void Xtoijk(const Real &x, const Real &y, const Real &z, int &i, int &j, int &k) const {
+    i = static_cast<int>(
+            std::floor((x - x_min_) / coords_.Dx(CoordinateDirection::X1DIR))) +
+        ib_s_;
+    j = (ndim_ > 1) ? static_cast<int>(std::floor(
+                          (y - y_min_) / coords_.Dx(CoordinateDirection::X2DIR))) +
+                          jb_s_
+                    : jb_s_;
+    k = (ndim_ > 2) ? static_cast<int>(std::floor(
+                          (z - z_min_) / coords_.Dx(CoordinateDirection::X3DIR))) +
+                          kb_s_
+                    : kb_s_;
+  }
+
+  // private:
+  int ib_s_;
+  int jb_s_;
+  int kb_s_;
+  Real x_min_;
+  Real x_max_;
+  Real y_min_;
+  Real y_max_;
+  Real z_min_;
+  Real z_max_;
+  Real x_min_global_;
+  Real x_max_global_;
+  Real y_min_global_;
+  Real y_max_global_;
+  Real z_min_global_;
+  Real z_max_global_;
```

Comment:
I was actually wondering if we need those or if the local ones are fine.
The boundary functions are only called for blocks that contain a boundary, so for those `x_max` is `x_max_global` and for all other blocks the boundary function the global numbers don't matter, do they?

Reply:
My reading is the boundary functions are called for all blocks regardless of whether there is a physical boundary. Either the boundary is physical or MPI, and we need to call the boundary functions after at least some `Receive`s if, for example, a particle is communicated from the X1R boundary periodically to the X1L boundary and needs its position updated from xmax + delta to xmin + delta. For that, these variables are explicitly used in `swarm_boundaries.hpp` e.g. `x = swarm_d.x_max_global_ - (swarm_d.x_min_global_ - x);`